lib/discourse_logstash_logger.rb

```diff
@@ -1,18 +1,103 @@
 # frozen_string_literal: true
 
-require "logstash-logger"
-
-class DiscourseLogstashLogger
-  def self.logger(uri:, type:)
-    LogStashLogger.new(
-      uri: uri,
-      sync: true,
-      customize_event: ->(event) do
-        event["severity_name"] = event["severity"]
-        event["severity"] = Object.const_get("Logger::Severity::#{event["severity"]}")
-        event["type"] = type
-        event["pid"] = Process.pid
-      end,
-    )
+require "json"
+require "socket"
+require_relative "git_utils"
+
+class DiscourseLogstashLogger < Logger
+  extend GitUtils
+
+  PROCESS_PID = Process.pid
+  HOST = Socket.gethostname
+
+  attr_accessor :customize_event, :type
+
+  # Creates a new logger instance.
+  #
+  # @param logdev [String, IO, nil] The log device. This can be one of:
+  #   - A string filepath: entries are written to the file at that path. If the file exists, new entries are appended.
+  #   - An IO stream (typically +$stdout+, +$stderr+, or an open file): entries are written to the given stream.
+  #   - nil or File::NULL: no entries are written.
+  # @param type [String] The type of log messages. This will add a `type` field to all log messages.
+  # @param customize_event [Proc, nil] A proc that customizes the log event before it is written to the log device.
+  #   The proc is called with a hash of log event data and can be modified in place.
+  #
+  # @return [Logger] A new logger instance with the specified log device and type.
+  def self.logger(logdev:, type:, customize_event: nil)
+    logger = self.new(logdev)
+    logger.type = type
+    logger.customize_event = customize_event if customize_event
+    logger
+  end
+
+  def add(*args, &block)
+    add_with_opts(*args, &block)
+  end
+
+  ALLOWED_HEADERS_FROM_ENV = %w[
+    REQUEST_URI
+    REQUEST_METHOD
+    HTTP_HOST
+    HTTP_USER_AGENT
+    HTTP_ACCEPT
+    HTTP_REFERER
+    HTTP_X_FORWARDED_FOR
+    HTTP_X_REAL_IP
+  ]
+
+  def add_with_opts(severity, message = nil, progname = nil, opts = {}, &block)
+    return true if @logdev.nil? || severity < @level
+
+    progname = @progname if progname.nil?
+
+    if message.nil?
+      if block_given?
+        message = yield
+      else
+        message = progname
+        progname = @progname
+      end
+    end
+
+    event = {
+      "message" => message,
+      "severity" => severity,
+      "severity_name" => Logger::SEV_LABEL[severity],
+      "pid" => PROCESS_PID,
+      "type" => @type.to_s,
+      "host" => HOST,
+      "git_version" => self.class.git_version,
+    }
+
+    # Only log backtrace and env for Logger::WARN and above.
+    # Backtrace is just noise for anything below that.
+    if severity >= Logger::WARN
+      if (backtrace = opts&.dig(:backtrace)).present?
+        event["backtrace"] = backtrace
+      end
+
+      if (env = opts&.dig(:env)).present?
+        ALLOWED_HEADERS_FROM_ENV.each do |header|
+          event["request.headers.#{header.downcase}"] = opts[:env][header]
+        end
+      end
```

Comment:
Note that the `env` and `backtrace` keys in opts are passed in by Logster to loggers that has been "chained" to it via the `Logster::Logger#chain` API. Basically, the instance of `Logster::Logger` will forward the logs to the loggers that have been chained. 

The change in https://github.com/discourse/logster/pull/238 updates `Logster::Logger` such that `Logster::Logger` will pass its `opts` to the chained logger if `with_opts: true` was set when calling `Logster::Logger#chain`